File: api/models/token.py
from datetime import datetime
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy import Column, String, JSON
from sqlalchemy.dialects.postgresql import JSONB

db = SQLAlchemy()

class Token(db.Model):
    __tablename__ = 'tokens'
    
    id = db.Column(db.Integer, primary_key=True)
    address = db.Column(db.String(42), unique=True, nullable=False)
    symbol = db.Column(db.String(255))
    name = db.Column(db.String(255))
    creator = db.Column(db.String(42))
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    website = db.Column(db.String(255))
    twitter = db.Column(db.String(255))
    telegram = db.Column(db.String(255))
    description = db.Column(db.Text)
    warpcast_url = db.Column(db.String(255))
    character_prompt = db.Column(db.Text)
    warpcast_app_key = db.Column(db.Text)
    ai_character = db.Column(JSON, nullable=True)
    twitter_oauth_token = db.Column(db.String(255))
    twitter_oauth_secret = db.Column(db.String(255))
    twitter_user_id = db.Column(db.String(255))
    twitter_username = db.Column(db.String(255))
    temp_request_token = db.Column(db.String(255))
    
    @classmethod
    def create_if_not_exists(cls, address):
        """Get existing token or create new one"""
        token = cls.query.filter_by(address=address.lower()).first()
        if not token:
            token = cls(
                address=address,
                creator=address,  # Default to self as creator initially
            )
            db.session.add(token)
            db.session.commit()
        return token

    def __init__(self, address, creator, **kwargs):
        self.address = address.lower()
        self.creator = creator.lower()
        for key, value in kwargs.items():
            setattr(self, key, value)

    def to_dict(self):
        return {
            'address': self.address,
            'creator': self.creator,
            'symbol': self.symbol,
            'name': self.name,
            'twitter': self.twitter,
            'telegram': self.telegram,
            'website': self.website,
            'description': self.description,
            'warpcast_url': self.warpcast_url,
            'character_prompt': self.character_prompt,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'ai_character': self.ai_character,
            'twitter_username': self.twitter_username,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None
        } 

File: services/app/models/token.js
import pool from './db';

export class Token {
  constructor(data) {
    this.address = data.address?.toLowerCase();
    this.symbol = data.symbol;
    this.name = data.name;
    this.creator = data.creator?.toLowerCase();
    this.created_at = data.created_at;
    this.website = data.website;
    this.twitter = data.twitter;
    this.telegram = data.telegram;
    this.description = data.description;
    this.warpcast_url = data.warpcast_url;
    this.character_prompt = data.character_prompt;
    this.warpcast_app_key = data.warpcast_app_key;
    this.ai_character = data.ai_character;
    this.twitter_oauth_token = data.twitter_oauth_token;
    this.twitter_oauth_secret = data.twitter_oauth_secret;
    this.twitter_user_id = data.twitter_user_id;
    this.twitter_username = data.twitter_username;
    this.temp_request_token = data.temp_request_token;
  }

  toDict() {
    return {
      address: this.address,
      creator: this.creator,
      symbol: this.symbol,
      name: this.name,
      twitter: this.twitter,
      telegram: this.telegram,
      website: this.website,
      description: this.description,
      warpcast_url: this.warpcast_url,
      character_prompt: this.character_prompt,
      created_at: this.created_at?.toISOString(),
      ai_character: this.ai_character,
      twitter_username: this.twitter_username,
    };
  }

  static async createIfNotExists(address) {
    const client = await pool.connect();
    try {
      // Check if token exists
      const existingResult = await client.query(
        'SELECT * FROM tokens WHERE address = $1',
        [address.toLowerCase()]
      );

      if (existingResult.rows.length > 0) {
        return new Token(existingResult.rows[0]);
      }

      // Create new token
      const result = await client.query(
        'INSERT INTO tokens (address, created_at) VALUES ($1, NOW()) RETURNING *',
        [address.toLowerCase()]
      );

      return new Token(result.rows[0]);
    } finally {
      client.release();
    }
  }

  static async findByAddress(address) {
    const result = await pool.query(
      'SELECT * FROM tokens WHERE address = $1',
      [address.toLowerCase()]
    );
    return result.rows.length ? new Token(result.rows[0]) : null;
  }

  async save() {
    const fields = [
      'symbol', 'name', 'creator', 'website', 'twitter', 'telegram',
      'description', 'warpcast_url', 'character_prompt', 'warpcast_app_key',
      'ai_character', 'twitter_oauth_token', 'twitter_oauth_secret',
      'twitter_user_id', 'twitter_username', 'temp_request_token'
    ];

    const setClause = fields
      .map((field, i) => `${field} = $${i + 2}`)
      .join(', ');
    
    const values = fields.map(field => this[field]);

    const query = `
      UPDATE tokens 
      SET ${setClause}
      WHERE address = $1
      RETURNING *
    `;

    const result = await pool.query(query, [this.address, ...values]);
    Object.assign(this, result.rows[0]);
    return this;
  }
} 